File: mixin-padang-webface/src/main/js/padang/functions/source/ReadCsv.ts
/*
 * Copyright (c) 2020-2023 Datamixin.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program. If not, see <http://www.gnu.org/licenses/>.
 */
import FunctionPlan from "padang/plan/FunctionPlan";
import ParameterPlan from "padang/plan/ParameterPlan";
import ParameterPlanUtils from "padang/plan/ParameterPlanUtils";
import FunctionPlanRegistry from "padang/plan/FunctionPlanRegistry";

import Function from "padang/functions/Function";

export default class ReadCsv extends Function {

	public static FUNCTION_NAME = "ReadCsv";

	constructor(
		public path: string,
		public delimiter: string,
		public firstRowHeader: boolean,
		public quoteCharacter: string,
	) {
		super(ReadCsv.FUNCTION_NAME);
	}

	public static PATH_PLAN: ParameterPlan = ParameterPlanUtils.createTextPlan(
		"path",
		"Path",
		"File or directory path"
	);
	public static DELIMITER_PLAN: ParameterPlan = ParameterPlanUtils.createTextPlan(
		"delimiter",
		"Value Delimiter",
		"Character separating values",
		",",
		"={Comma: ',', Colon: ':', Tab: '\\t', Space:' '}"
	);
	public static FIRST_ROW_HEADER_PLAN: ParameterPlan = ParameterPlanUtils.createLogicalPlan(
		"firstRowHeader",
		"First Row Header",
		"First row is a header (column names)",
		true
	);
	public static QUOTE_CHARACTER_PLAN: ParameterPlan = ParameterPlanUtils.createTextPlan(
		"quoteCharacter",
		"Quote Character",
		"Single character as start and end quoted value",
		'"',
		"={DoubleQuote: '\"', SingleQuote: '\\''}"
	);
	public static PARSE_DATES_PLAN: ParameterPlan = ParameterPlanUtils.createTextPlan(
		"parseDates",
		"Parse Dates",
		"Parse date columns",
		"=[]"
	);

	public static getPlan(): FunctionPlan {
		let plan = new FunctionPlan(
			ReadCsv.FUNCTION_NAME,
			"Read CSV",
			"mdi-file-delimited-outline",
			"Create dataset from CSV File"
		);
		let list = plan.getParameterList();
		list.add(ReadCsv.PATH_PLAN);
		list.add(ReadCsv.DELIMITER_PLAN);
		list.add(ReadCsv.FIRST_ROW_HEADER_PLAN);
		list.add(ReadCsv.QUOTE_CHARACTER_PLAN);
		list.add(ReadCsv.PARSE_DATES_PLAN);
		return plan;
	}

}

let registry = FunctionPlanRegistry.getInstance();
registry.registerPlan(ReadCsv.getPlan());

File: mixin-padang-garut/src/main/python/garut/functions/source/read_csv.py
#
# Copyright (c) 2020-2023 Datamixin.
#
# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program. If not, see <http://www.gnu.org/licenses/>.
#
import pandas as pd
from typing import Dict
from garut.contexts.context import Context
from garut.instruction import InstructionFactory
from garut.functions.function import Function, FunctionRegistry
from garut.functions.dataset.dataset_function import DatasetFunction


class ReadCsv(Function):

    FUNCTION_NAME = "ReadCsv"

    PATH = "path"
    DELIMITER = "delimiter"
    FIRST_ROW_HEADER = "firstRowHeader"
    QUOTE_CHARACTER = "quoteCharacter"
    PARSE_DATES = "parseDates"

    def execute(self, context: Context, options: Dict[str, any]) -> any:
        path = options.get(ReadCsv.PATH, None)
        if path is None or path == "":
            dataFrame = pd.DataFrame()
        else:
            header = self._getHeader(options)
            dataFrame = pd.read_csv(
                path,
                sep=options.get(ReadCsv.DELIMITER, ","),
                header=header,
                quotechar=options.get(ReadCsv.QUOTE_CHARACTER, '"'),
                parse_dates=options.get(ReadCsv.PARSE_DATES, False),
            )

            # Header if header=None
            if header is None:
                dataFrame.columns = ["Column" + str(index) for index in dataFrame.columns.values]

            # Rubah column ber tipe object menjadi string
            dataFrame = DatasetFunction.forceObjectColumnsToString(dataFrame)
        return dataFrame

    def _getHeader(self, options: Dict[str, any]) -> str:
        firstRowHeader = options.get(ReadCsv.FIRST_ROW_HEADER, True)
        if firstRowHeader:
            return 0
        else:
            return None


parameters = [ReadCsv.PATH, ReadCsv.DELIMITER, ReadCsv.FIRST_ROW_HEADER, ReadCsv.QUOTE_CHARACTER, ReadCsv.PARSE_DATES]
registry: FunctionRegistry = FunctionRegistry.getInstance()
registry.register(ReadCsv.FUNCTION_NAME, ReadCsv, parameters)

factory: InstructionFactory = InstructionFactory.getInstance()
factory.registerInstruction(ReadCsv.FUNCTION_NAME)
